### ISSUE TITLE
[RKE] CIS Benchmark 1.24 requires file permissions to be 600 instead of 644

### ISSUE THREAD

**Issue by andypitcher** [rancher/rancher#42012]
@snasovich I'm creating this issue to improve the file permissions of `RKE` files based on CIS new requirements.

Here is the list of concerned CIS check along with the files that would need to be set to 600 instead of 644.

[4.1.7](https://github.com/rancher/security-scan/blob/687e2b3556db4b295918b907de059f247d7c16a6/package/cfg/rke-cis-1.24-hardened/node.yaml#L104) hardened: /etc/kubernetes/ssl/kube-ca.pem

Related issue: https://github.com/rancher/rancher/issues/41131#issuecomment-1603464339

@rayandas could you confirm is the only location ?

**Comment by rayandas** [comment]
@andypitcher yes /etc/kubernetes/ssl/kube-ca.pem is the only location for RKE.

**Comment by vardhaman22** [comment]
this check fails on newer cis profiles (1.7, 1.8) as well

**Comment by vardhaman22** [comment]
I tried to make cis 4.1.7 test case pass for rke hardened cluster, currently it passes for controlplane and worker nodes but fails for etcd node, and here are my findings around that:
1. cis 4.1.7 check needs permission for kube-ca.pem file to be 600 and 4.1.8 check needs root to be the owner of the file.
2. since in rke cluster etcd also uses kube-ca.pem file (for argument: --trusted-ca-file) it needs read access to the file. If we set permission of this file to 600 and owner as root the etcd container fails to start.
3. if we make etcd user the owner of the file then 4.1.8 cis check will fail.
so this can't be fixed by simply changing that file's permission to 600. I think one solution is we use different ca certificate file for etcd may be with the same content as kube-ca.pem. But this change is only needed in the cluster to make cis checks pass for hardened cluster other than that functionality wise there is no effect of this.
**Note: for non hardened cluster user of etcd container  is root. But for hardened cluster we create an etcd user and that user is used for etcd container.**

**Comment by snasovich** [comment]
Due to larger than expected impact of the change and additional testing needed, pushing it into the next milestone.

**Comment by vardhaman22** [comment]
Tried the following solution ([PR](https://github.com/rancher/rke/pull/3543))

1. added a new etcd CA cert which is used to create certificates needed for etcd i.e. etcd server certificates named as kube-etcd-{host-IP}
2. introduce a new etcd certificate kube-apiserver-client certificate which is used by kube api server to connect to etcd.
We will also use this certificate to check if etcd is healthy after etcd plan deployment and etcd reconciliation. earlier the KubeNodeCert was used.
 - https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/services/etcd.go#L92
 - https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/cluster/reconcile.go#L175
3. for etcd snapshot and restore used the server certificates. 
	Right now mostly KubeNodeCert is used which is the cert needed for kubelet 
	- https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/services/etcd.go#L402
	- https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/services/etcd.go#L912
	etc.
	but in some places kube-etcd-{host-IP} cert is also used.
	- https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/services/etcd.go#L659

So decided not to introduce a new certificates for etcd restore.

4. RKE will pass an ENV variable to RKE_TOOLS (cert downloader and cert deployer) to specify that kube-ca.pem file permissions needs to be
updated or not.
 - https://github.com/rancher/rke-tools/blob/6fd84ca5995e8fb26ff940b9fccc002800885c0e/cert-deployer#L41
 - https://github.com/rancher/rke/blob/3b5fd33f25a678b072049a13bc64da526b18d961/pki/deploy.go#L402

This is needed so that same RKE_TOOLS image can be used with the OOB released versions, basically compatibility with old rke.


**And did following tests:**

1. Cluster Provisioning with old rke-tools image. v1.27.11-rancher1-1 with new rancher
--- Successful
2. created a deployment, took local snapshot and restore (with etcd + kubernetes + config option).
--- Successful
3. taking a backup.
--- Successful
4. adding more etcd nodes (scale from 1 node to 3 node)
--- Successful.
5. deleted deployment and tried restoring snapshot.
--- snapshot restored successfully
6. upgrading to a k8s version with new custom rke-tools image v1.28.7-rancher1-2
--- successful
7. deleted deployment and tried restoring from snapshot with cluster config, kubernetes version and etcd option.
---- successful
8. rotate certificate from rancher
	1. all certificates: successful
	2. rotate individual service:
		etcd: successful
		kubelet: successful
		kube-api-server: successful
		kube-controller-manager: successful
9. restore from the initially create snapshot
---- successful
10. deleting a cluster
---- successful


11. created 1.28 cluster with v2.8.3 rancher, created a deployment and took snapshot of the cluster .
12. upgraded rancher to the newly created custom rancher image.
 -- created deployment in the cluster(which still has old certificates i.e. no etcd CA)
 -- deployed CIS benchmark chart.
 -- Tried restoring snapshot which was created with old rancher
 **FAILED**
**since etcd snapshot restore tries to find etcd CA certificate which is used for container arguments. And it is not able to find it in the cluster that's why it failed.
So to avoid this, this change should only be introduced with a new kubernetes version and rancher version..... so that cluster with any kubernetes versions prior to that version keeps functioning in the same way, also it will protect OOB releases of that version.**

**Comment by andypitcher** [comment]
> I tried to make cis 4.1.7 test case pass for rke hardened cluster, currently it passes for controlplane and worker nodes but fails for etcd node, and here are my findings around that:
> 
> 1. cis 4.1.7 check needs permission for kube-ca.pem file to be 600 and 4.1.8 check needs root to be the owner of the file.
> 2. since in rke cluster etcd also uses kube-ca.pem file (for argument: --trusted-ca-file) it needs read access to the file. If we set permission of this file to 600 and owner as root the etcd container fails to start.
> 3. if we make etcd user the owner of the file then 4.1.8 cis check will fail.
>    so this can't be fixed by simply changing that file's permission to 600. I think one solution is we use different ca certificate file for etcd may be with the same content as kube-ca.pem. But this change is only needed in the cluster to make cis checks pass for hardened cluster other than that functionality wise there is no effect of this.
>    **Note: for non hardened cluster user of etcd container  is root. But for hardened cluster we create an etcd user and that user is used for etcd container.**

@vardhaman22 thanks for going through these tests, it doesn't seem trivial to comply to the defaults of CIS.
This said, the goal of CIS is to set a security baseline for most of the k8s distributions, but sometimes we can define our own tests and expected outcome, if we can't comply. As long as we provide a justification for it. This is why most of the k8s distro have their own CIS profiles, as much as we do.

Based on your two last comments, and given the difficulty we might encounter to align with the permission modification of RKE's kube-ca.pem from 644 to 600, here is what I would propose:

- Lower the permissions of `kube-ca.pem` from `600` to 640.
- Change the ownership of `kube-ca.pem` from `root:root` to `root:etcd`.

My assumption is that the `etcd` group only holds the `etcd` user, which is by itself a restricted system user that can access highly critical resources. In this context, compromising the `etcd` group/user already poses a higher security impact than gaining access to kube-ca.pem by any user (644).
Overall assigning the `etcd` group to kube-ca.pem helps maintain a balance between security and functionality, ensuring that only authorized processes related to etcd management (such as snapshot) can read this file.
Finally, we can update 4.1.7 and 4.1.8 tests accordingly, along with the remediation and docs to reflect our decision.

**Comment by vardhaman22** [comment]
@andypitcher i am trying out a solution for the above problem in a [PR](https://github.com/rancher/rke/pull/3562) to have different etcd CA cert in the cluster for k8s version >=1.29 and i am trying to test it, it is not ready yet since i encounterd some failing scenarios regarding snapshot restore and cluster upgrades.
the approach to change the permission of the files as you suggested will be much safer. If we can proceed with it that would be great.
@snasovich @kinarashah please share your opinions on this.
